internal/handlers/commonparams/extract_params.go

```diff
@@ -0,0 +1,250 @@
+			panic(err)
+		}
+
+		if options == nil {
+			return lazyerrors.Errorf("unexpected field '%s' encountered", lookup)
+		}
+
+		if options.ignored {
+			l.Debug(
+				"ignoring field",
+				zap.String("command", doc.Command()), zap.String("field", key), zap.Any("value", val),
+			)
+
+			continue
+		}
+
+		if options.unimplemented {
+			msg := fmt.Sprintf(
+				"%s: support for field %q with value %v is not implemented yet",
+				doc.Command(), key, val,
+			)
+
+			return commonerrors.NewCommandErrorMsgWithArgument(commonerrors.ErrNotImplemented, msg, key)
+		}
+
+		if options.nonDefault {
+			v, ok := val.(bool)
+
+			if ok && v {
+				msg := fmt.Sprintf(
+					"%s: support for field %q with non-default value %v is not implemented yet",
+					doc.Command(), key, val,
+				)
+
+				return commonerrors.NewCommandErrorMsgWithArgument(commonerrors.ErrNotImplemented, msg, key)
+			}
+		}
+
+		err = setStructField(&elem, fieldIndex, command, key, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// tagOptions contains options for the structure field tag.
+type tagOptions struct {
+	optional      bool
+	nonDefault    bool
+	unimplemented bool
+	ignored       bool
+}
+
+// lookupFieldTag looks for the tag and returns its options.
+func lookupFieldTag(key string, value *reflect.Value) (int, *tagOptions, error) {
+	var to tagOptions
+	var i int
+	var found bool
+
+	for ; i < value.NumField(); i++ {
+		field := value.Type().Field(i)
+
+		tag := field.Tag.Get("ferretdb")
+
+		optionsList := strings.Split(tag, ",")
+
+		if len(optionsList) == 0 {
+			return 0, nil, lazyerrors.Errorf("no tag provided for %s", field.Name)
+		}
+
+		if optionsList[0] != key {
+			continue
+		}
+
+		for _, tt := range optionsList[1:] {
+			switch tt {
+			case "opt":
+				to.optional = true
+			case "non-default":
+				to.nonDefault = true
+			case "unimplemented":
+				to.unimplemented = true
+			case "ignored":
+				to.ignored = true
+			default:
+				return 0, nil, lazyerrors.Errorf("unknown tag option %s", tt)
+			}
+		}
+
+		found = true
+
+		break
+	}
+
+	if !found {
+		return 0, nil, nil
+	}
+
+	return i, &to, nil
+}
+
+// setStructField sets the value of the document field to the structure field.
+func setStructField(elem *reflect.Value, i int, command, key string, val any) error {
+	var err error
+
+	field := elem.Type().Field(i)
+
+	// Set the value of the field from the document.
+	fv := elem.Field(i)
+	if !fv.CanSet() {
+		panic(fmt.Sprintf("field %s is not settable", field.Name))
+	}
+
+	var settable any
+
+	switch fv.Kind() { //nolint: exhaustive // all other types are not supported
+	case reflect.Int32, reflect.Int64, reflect.Float64:
+		if key == "skip" || key == "limit" {
+			settable, err = GetWholeParamStrict(command, key, val)
+			if err != nil {
+				return err
+			}
+
+			break
+		}
+
+		settable, err = GetWholeNumberParam(val)
+		if err != nil {
+			return err
+		}
+	case reflect.String, reflect.Bool, reflect.Struct, reflect.Pointer:
```

Comment:
Could it work on parameter like below array? :thinking: https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/pg/msg_aggregate.go#L85

Reply:
Yes, it should work for `*types.Array` as it's a pointer to structure. It's checked by `TestCountCompatCommandErrors/QueryArray`.

Reply:
Perfect